src/server/src/uio/connection.rs

```
@@ -0,0 +1,341 @@
+    io: IoDriver,
+    db: Db,
+    rbuf: ReadBuf,
+    wbuf: WriteBuf,
+    num_inflights: usize,
+}
+
+impl Connection {
+    pub fn new(id: u64, fd: RawFd, io: IoDriver, db: Db) -> Connection {
+        let mut conn = Self {
+            id,
+            fd,
+            io,
+            db,
+            rbuf: ReadBuf::new(),
+            wbuf: WriteBuf::new(),
+            num_inflights: 0,
+        };
+        conn.prepare_read();
+        conn
+    }
+
+    // Handles a completion event.
+    //
+    // Returns true if the connection should be dropped.
+    pub fn tick(&mut self, cqe: cqueue::Entry) -> bool {
+        self.num_inflights -= 1;
+
+        let op = Token::op(cqe.user_data());
+        let result = check_io_result(cqe.result());
+        match op {
+            opcode::Read::CODE => match result {
+                Ok(size) => {
+                    trace!(self.id, size, "connection complete read");
+                    self.complete_read(size as usize);
+                }
+                Err(err) => error!(%err, self.id, "connection complete read"),
+            },
+            opcode::Write::CODE => match result {
+                Ok(size) => {
+                    trace!(self.id, size, "connection complete write");
+                    self.complete_write(size as usize);
+                }
+                Err(err) => error!(%err, self.id, "connection complete write"),
+            },
+            opcode::Close::CODE => match result {
+                Ok(_) => trace!(self.id, "connection complete close"),
+                Err(err) => error!(%err, self.id, "connection complete close"),
+            },
+            _ => unreachable!(),
+        }
+
+        self.num_inflights == 0
+    }
+
+    fn process(&mut self) {
+        while let Some(frame) = self.rbuf.parse_frame() {
+            let cmd = Command::from_frame(frame).unwrap();
+            let reply = cmd.apply(&self.db).unwrap();
+            self.wbuf.write_frame(&reply);
```

@huachaohuang we may not call `unwrap` here since we don't support many commands now. If we call `unwrap`, instead of returning client an Error response, it will panic.

---

My mistake. The panic I meet in local env isn't caused by these lines.